src/sleep.rs

```diff
@@ -0,0 +1,110 @@
+//! Driver for light and deep sleep.
+//!
+//! Currently implemented Timer and UART wakeup sources.
+
+use esp_idf_sys::*;
+use std::time::Duration;
+
+#[derive(Debug)]
+pub enum WakeupSource {
+    Timer { dur: Duration },
+    Uart { uart_num: i32, threshold: i32 },
+    // TODO: add more sources
+}
+
+impl WakeupSource {
+    pub fn apply(&self) -> Result<(), EspError> {
+        match *self {
+            WakeupSource::Timer { dur } => {
+                esp!(unsafe { esp_sleep_enable_timer_wakeup(dur.as_micros() as u64) })?;
+            }
+            WakeupSource::Uart {
+                uart_num,
+                threshold,
+            } => {
+                esp!(unsafe { uart_set_wakeup_threshold(uart_num, threshold) })?;
+                esp!(unsafe { esp_sleep_enable_uart_wakeup(uart_num) })?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub trait Sleep {
+    fn add_wakeup_source(&mut self, wakeup: WakeupSource) -> Result<(), EspError>;
+    fn sleep(&self) -> Result<(), EspError>;
+
+    fn apply_wakeup_sources(&self, wakeup_sources: &[WakeupSource]) -> Result<(), EspError> {
+        // reset wakeup sources
+        esp!(unsafe { esp_sleep_disable_wakeup_source(esp_sleep_source_t_ESP_SLEEP_WAKEUP_ALL) })?;
+
+        for wakeup in wakeup_sources.iter() {
+            wakeup.apply()?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct LightSleep {
+    wakeup_sources: Vec<WakeupSource>,
```

Comment:
It is better if this one is `heapless::Vec` bound by the number of wakeup sources, which I would assume is small and finite. Not worth it to require the `alloc` (or let alone `std`) functionality just for this.

Yet, given that I suggest a radical simplification of the API and checking a major precondition first - that sleep sources between light and deep sleep API are the same - which I doubt - this is irrelevant.

Reply:
well, there is a limitied number of wakeup sources, but the External Wakeup (ext1) source can be configured so multiple RTC pins can be used to wake up. So either there must be possible to define a variable number of WakeupSource::Ext1, or one WakeupSource::Ext1 must have a list of pins.

But lets put this abit on hold to see how stuff plays out in the other comment threads

Reply:
But both Ext1 and Ext2 are each a **single** data source - at least under deep sleep - where the pins you pass are AND-ed or OR-ed together. And by the way the pins can be kept in a bitset or suchlike. What I'm saying is that using an allocated vector probably still cannot be justified by referring to Ext1 or Ext2.